### PR TITLE
ci: fix ovn-vpc-nat-gw e2e failed

### PR DESCRIPTION
**Pull request by zcq98**
# Pull Request

- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Tests

## Which issue(s) this PR fixes

[ovn-vpc-nat-gw e2e failed](https://github.com/kubeovn/kube-ovn/actions/runs/7122458767/job/19394449817?pr=3500)

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 5d1aafd</samp>

Enhanced the e2e test for ovn-vpc-nat-gw to cover multiple subnets in the underlay network. Updated `test/e2e/ovn-vpc-nat-gw/e2e_test.go` to handle the additional resources.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 5d1aafd</samp>

> _`vlan` and `subnet`_
> _added to test underlay_
> _autumn network leaves_

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 5d1aafd</samp>

* Add a variable `vlanExtraName` to store the name of an extra vlan for testing multiple subnets ([link](https://github.com/kubeovn/kube-ovn/pull/3503/files?diff=unified&w=0#diff-c1831d6aa37f96132386404bff7ef6c738c40bcef66ba7492aca28929d61dca2L84-R84))
* Initialize `vlanExtraName` with a random suffix to avoid name conflicts ([link](https://github.com/kubeovn/kube-ovn/pull/3503/files?diff=unified&w=0#diff-c1831d6aa37f96132386404bff7ef6c738c40bcef66ba7492aca28929d61dca2R161))
* Create an extra vlan with `vlanExtraName` and `providerExtraNetworkName` ([link](https://github.com/kubeovn/kube-ovn/pull/3503/files?diff=unified&w=0#diff-c1831d6aa37f96132386404bff7ef6c738c40bcef66ba7492aca28929d61dca2R698-R701))
* Create a subnet for the extra vlan with `underlayExtraSubnetName`, `vlanExtraName`, and the cidr and gateway from `extracidr` and `extragateway` ([link](https://github.com/kubeovn/kube-ovn/pull/3503/files?diff=unified&w=0#diff-c1831d6aa37f96132386404bff7ef6c738c40bcef66ba7492aca28929d61dca2L725-R736))
* Check that the subnet has the expected vlan name and cidr block ([link](https://github.com/kubeovn/kube-ovn/pull/3503/files?diff=unified&w=0#diff-c1831d6aa37f96132386404bff7ef6c738c40bcef66ba7492aca28929d61dca2L725-R736))
* Delete the extra vlan after the test is done to clean up the resources ([link](https://github.com/kubeovn/kube-ovn/pull/3503/files?diff=unified&w=0#diff-c1831d6aa37f96132386404bff7ef6c738c40bcef66ba7492aca28929d61dca2R473-R474))

### CODE REVIEW

**Comment by zcq98** [comment]
hi @bobz965, could you help review this ?

**Review by bobz965 — APPROVED** [review]

**Comment by bobz965** [comment]
`[FAILED] wait for subnet "external" to disappear`  external subnet 清理过程不稳定，偶现 delete 超时

**Comment by zcq98** [comment]
image
偶发性地add finalizer失败 导致subnet无法被清除

**Comment by bobz965** [comment]
> image 偶发性地add finalizer失败 导致subnet无法被清除

这个finalizer的update接口改为使用patch应该可以解决偶发性不能更新成功的情况。

File: pkg/controller/subnet.go [hunk 1]
```diff
@@ -479,8 +479,15 @@ func checkAndUpdateExcludeIPs(subnet *kubeovnv1.Subnet) bool {
 
 func (c *Controller) handleSubnetFinalizer(subnet *kubeovnv1.Subnet) (bool, error) {
 	if subnet.DeletionTimestamp.IsZero() && !util.ContainsString(subnet.Finalizers, util.ControllerName) {
-		subnet.Finalizers = append(subnet.Finalizers, util.ControllerName)
-		if _, err := c.config.KubeOvnClient.KubeovnV1().Subnets().Update(context.Background(), subnet, metav1.UpdateOptions{}); err != nil {
+		newSubnet := subnet.DeepCopy()
```
Comment (bobz965) [inline pkg/controller/subnet.go:482]: 之前已经被deepcopy过了，不需要再copy一次。

Comment (zcq98) [reply pkg/controller/subnet.go:482]: deepcopy还是有必要的，否则kube-ovn-conformance-e2e过不了

**Review by oilbeater — APPROVED** [review]

**Pull request merged**